### PR TITLE
fix(merge): better handling of branch protection

### PR DESCRIPTION
**Pull request by sileht**
This change waits only if Branch Protection are blocked on "Required
Check" message.

### CODE REVIEW

File: mergify_engine/actions/merge/action.py [hunk 1]
```diff
@@ -365,3 +383,9 @@ def _handle_merge_error(self, e, ctxt):
                 error_message=e.message,
             )
             return ("failure", message, f"GitHub error message: `{e.message}`")
+
+    @staticmethod
+    def _extract_required_check(message):
+        m = REQUIRED_STATUS_RE.match(message)
+        if m:
+            return m.group(1)
```
Comment (jd) [inline mergify_engine/actions/merge/action.py:391]:
```suggestion
        if m := REQUIRED_STATUS_RE.match(message):
            return m.group(1)
```

File: mergify_engine/actions/merge/action.py [hunk 2]
```diff
@@ -343,19 +344,36 @@ def _handle_merge_error(self, e, ctxt):
...
-            return (
-                None,
-                "Waiting for the Branch Protection to be validated",
-                "Branch Protection is enabled and is preventing Mergify "
-                "to merge the pull request. Mergify will merge when "
-                "branch protection settings validate the pull request. "
-                f"(detail: {e.message})",
-            )
+            check_name = self._extract_required_check(e.message)
+            if check_name:
```
Comment (jd) [inline mergify_engine/actions/merge/action.py:348]: This variable is actually unused. You might want to at least log it to be sure it's correct?

File: mergify_engine/actions/merge/action.py [hunk 3]
```diff
@@ -343,19 +344,36 @@ def _handle_merge_error(self, e, ctxt):
...
+                    None,
+                    "Waiting for the Branch Protection Required Checks to be validated",
+                    "Branch Protection is enabled and is preventing Mergify "
+                    "to merge the pull request. Mergify will merge when "
+                    "branch protection settings validate the pull request. "
+                    f"(detail: {e.message})",
+                )
+            else:
+                ctxt.log.info(
+                    "The Branch Protection is not validated anymore",
```
Comment (jd) [inline mergify_engine/actions/merge/action.py:364]:
```suggestion
                    "Branch protection settings are not validated anymore",
```

File: mergify_engine/actions/merge/action.py [hunk 4]
```diff
@@ -343,19 +344,36 @@ def _handle_merge_error(self, e, ctxt):
...
+                    status=e.status_code,
+                    error_message=e.message,
+                )
+
+                return (
+                    "cancelled",
+                    "The Branch Protection is not validated anymore",
+                    "Branch Protection is enabled and is preventing Mergify "
```
Comment (jd) [inline mergify_engine/actions/merge/action.py:372]:
```suggestion
                    "[Branch protection](https://docs.github.com/en/github/administering-a-repository/about-protected-branches) is enabled and is preventing Mergify "
```

File: mergify_engine/actions/merge/action.py [hunk 5]
```diff
@@ -343,19 +344,36 @@ def _handle_merge_error(self, e, ctxt):
...
+            check_name = self._extract_required_check(e.message)
+            if check_name:
+                ctxt.log.info(
+                    "Waiting for the Branch Protection Required Checks to be validated",
+                    status=e.status_code,
+                    error_message=e.message,
+                )
+                return (
+                    None,
+                    "Waiting for the Branch Protection Required Checks to be validated",
```
Comment (jd) [inline mergify_engine/actions/merge/action.py:356]:
```suggestion
                    "Waiting for the branch protection required status checks to be validated",
```

File: mergify_engine/actions/merge/action.py [hunk 6]
```diff
@@ -343,19 +344,36 @@ def _handle_merge_error(self, e, ctxt):
...
+                    "Branch Protection is enabled and is preventing Mergify "
```
Comment (jd) [inline mergify_engine/actions/merge/action.py:357]:
```suggestion
                    "[Branch protection](https://docs.github.com/en/github/administering-a-repository/about-protected-branches) is enabled and is preventing Mergify "
```

File: mergify_engine/actions/merge/action.py [hunk 7]
```diff
@@ -343,19 +344,36 @@ def _handle_merge_error(self, e, ctxt):
...
+                    "to merge the pull request. Mergify will merge when "
+                    "branch protection settings validate the pull request. "
+                    f"(detail: {e.message})",
```
Comment (jd) [inline mergify_engine/actions/merge/action.py:360]:
```suggestion
                    "to merge the pull request. Mergify will merge when "
                    "the [required status check](https://docs.github.com/en/github/administering-a-repository/about-required-status-checks) validate the pull request. "
                    f"(detail: {e.message})",
```

File: mergify_engine/actions/merge/action.py [hunk 8]
```diff
@@ -343,19 +344,36 @@ def _handle_merge_error(self, e, ctxt):
...
-                "Waiting for the Branch Protection to be validated",
-                "Branch Protection is enabled and is preventing Mergify "
-                "to merge the pull request. Mergify will merge when "
-                "branch protection settings validate the pull request. "
-                f"(detail: {e.message})",
-            )
+            check_name = self._extract_required_check(e.message)
+            if check_name:
+                ctxt.log.info(
+                    "Waiting for the Branch Protection Required Checks to be validated",
```
Comment (jd) [inline mergify_engine/actions/merge/action.py:350]:
```suggestion
                    "Waiting for the branch protection required status checks to be validated",
```

File: mergify_engine/actions/merge/action.py [hunk 9]
```diff
@@ -343,19 +344,36 @@ def _handle_merge_error(self, e, ctxt):
...
+                    status=e.status_code,
+                    error_message=e.message,
+                )
+
+                return (
+                    "cancelled",
+                    "The Branch Protection is not validated anymore",
+                    "Branch Protection is enabled and is preventing Mergify "
+                    "to merge the pull request. Mergify will merge when "
+                    "branch protection settings validate the pull request again. "
```
Comment (jd) [inline mergify_engine/actions/merge/action.py:374]:
```suggestion
                    "branch protection settings validate the pull request once again. "
```

File: mergify_engine/tests/functional/test_engine.py [hunk 10]
```diff
@@ -1223,7 +1223,7 @@ def test_merge_branch_protection_ci(self):
         assert checks[0]["conclusion"] is None
         assert "in_progress" == checks[0]["status"]
         assert (
-            "Waiting for the Branch Protection to be validated"
+            "Waiting for the Branch Protection Required Checks to be validated"
```
Comment (jd) [inline mergify_engine/tests/functional/test_engine.py:1226]:
```suggestion
            "Waiting for the branch protection required status checks to be validated"
```

**Review by jd — CHANGES_REQUESTED** [review]
This is awesome. I've suggested a few rewrite and links to the GitHub doc to avoid support request with users not at ease with Mergify. <3

**Review by jd — CHANGES_REQUESTED** [review]
Looks like some were missed?
I don't care so much about logs, but the link might be worth adding.

File: mergify_engine/actions/merge/action.py (hunk 2 above)
Comment (sileht) [reply mergify_engine/actions/merge/action.py]: We already log the e.message and put it in the summary. We don't need this variable.

**Comment by sileht** [comment]
I though I fixed all of them...

**Review by jd — APPROVED** [review]

**Pull request merged**